src/engine/enginebuffer.cpp

```diff
@@ -525,6 +530,17 @@ void EngineBuffer::slotTrackLoaded(TrackPointer pTrack,
     // Reset the pitch value for the new track.
     m_pause.unlock();
 
+    while (!m_activeMacro.empty()) {
+        m_activeMacro.pop();
```

Comment:
A single thread is only allowed to either act as a *consumer* or *producer* of a lock-free queue. Invoking both push() and pop() from the same thread is wrong and may cause undefined behavior.

Reply:
May I ask where that information is from? The only related documentation I could find was this:
> Only a single writer thread can perform enqueue operations and only a single reader thread can perform dequeue operations. Any other usage is invalid.
Which in my eyes doesn't seem to exclude the possibility of both being on the same thread.

In essence I need a queue here from which I can pop without reallocation. Alternatively I could use a simple list with a separate index counter since this creation is (to my understanding) not happening in the real-time thread.
Though that could change if live looping is implemented, where the Macro is immediately queued after recording.

Reply:
The synchronization with memory-barriers is only effective if enqueue and dequeue operations are strictly performed by the *same* threads respectively. Alternating between enqueuing and dequeuing within a thread is not permitted. If you want to know why then read the implementation.

If both of these threads are identical you wouldn't need a lock-free queue here ;)

PS: In Rust this unsafe C++ code wouldn't even be possible, because you only have access to either the *producer* or *consumer* end of the queue in each thread, not both.